### ISSUE TITLE
crossv_kfold, crossing, and mutate lead to a segmentation fault

### ISSUE THREAD

**Issue by dgrtwo** [tidyverse/dplyr#2680]
`mutate` leads to segmentation faults on `resample` objects from the modelr package when those `resample` objects have been repeated in a list. This is particularly noticeable with tidyr's `crossing()`, which is how I typically try multiple tuning parameters in modelr.

```r
library(modelr)
library(dplyr)
library(tidyr)
library(purrr)

# minimal example- we're not actually using param, but it
# could be a machine learning tuning parameter
crosses <- mtcars %>%
  crossv_kfold(k = 5) %>%
  crossing(param = 1:5)

# WARNING: this next line causes a segfault for me:
crosses <- crosses %>%
  mutate(model = map(train, ~ lm(mpg ~ wt, .)))
```

Replacing `map` with `lapply`, or `crossing` with a manual version using indexing, leads to the same issue. Similarly, approaches that avoid using `mutate` work fine, or that only do something trivial with the training data like turning it into a data frame rather than fitting a linear model:

```r
# each of these lines approaches work
crosses$model <- map(crosses$train, ~ lm(mpg ~ wt + disp, .))
crosses$model <- with(crosses, map(train, ~ lm(mpg ~ wt + disp, .)))

# so does doing something simple
crosses %>%
  mutate(d = map(train, as.data.frame))
```

I suppose this probably has something to do with the fact that `resample` objects have "pointers" to the same data, but I'm not sure how that could interact with `mutate` and the environment to cause a segmentation fault.

**Comment by hadley** [comment]
Is this with the dev version?

**Comment by dgrtwo** [comment]
Yes

**Comment by dgrtwo** [comment]
Ah: I just confirmed the bug doesn't happen with CRAN dplyr (0.5.0)

**Comment by krlmlr** [comment]
What version of bindrcpp are you using?

**Comment by krlmlr** [comment]
Never mind, I forgot that we fixed the pointer leak in dplyr.

**Comment by krlmlr** [comment]
I can't replicate the crash on Ubuntu. Could you please post `session_info()`?

**Comment by dgrtwo** [comment]
Shown below, thanks. Note that in some rare cases it didn't crash until I printed it, or more rarely it resulted in `train` and `test` columns being filled with nonsense data from elsewhere in my namespace (classic memory bug issues)

```
> devtools::session_info()
Session info -----------------------------------------------------------------------------------------------------
 setting  value                       
 version  R version 3.3.3 (2017-03-06)
 system   x86_64, darwin13.4.0        
 ui       RStudio (1.0.136)           
 language (EN)                        
 collate  en_US.UTF-8                 
 tz       America/New_York            
 date     2017-04-17                  

Packages ---------------------------------------------------------------------------------------------------------
 package    * version    date       source                          
 assertthat   0.2.0      2017-04-11 cran (@0.2.0)                   
 devtools     1.12.0     2016-06-24 CRAN (R 3.3.0)                  
 digest       0.6.12     2017-01-27 cran (@0.6.12)                  
 dplyr      * 0.5.0.9002 2017-04-17 Github (tidyverse/dplyr@3eb9429)
 glue         1.0.0      2017-04-17 Github (tidyverse/glue@3791aef) 
 magrittr     1.5        2014-11-22 CRAN (R 3.3.0)                  
 memoise      1.0.0      2016-01-29 CRAN (R 3.3.0)                  
 modelr     * 0.0.0.9000 2017-04-17 local                           
 purrr      * 0.2.2      2016-06-18 cran (@0.2.2)                   
 R6           2.2.0      2016-10-05 cran (@2.2.0)                   
 Rcpp         0.12.10.1  2017-04-17 Github (RcppCore/Rcpp@052089b)  
 rlang        0.0.0.9017 2017-04-17 Github (hadley/rlang@8fd2651)   
 tibble       1.3.0      2017-04-01 cran (@1.3.0)                   
 tidyr      * 0.6.1      2017-01-10 CRAN (R 3.3.2)                  
 withr        1.0.2      2016-06-20 cran (@1.0.2)                   

```

**Comment by hadley** [comment]
@krlmlr I'm also seeing a random segfault: https://travis-ci.org/tidyverse/dplyr/builds/222951840. Possibly related?

**Comment by krlmlr** [comment]
Seeing errors with `--valgrind` and `gctorture()`. @dgrtwo: Is your code supposed to complete without error?

```
> R -q --debugger=valgrind -f clion-test.R 
==1435== Memcheck, a memory error detector
==1435== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==1435== Using Valgrind-3.12.0.SVN and LibVEX; rerun with -h for copyright info
==1435== Command: /usr/lib/R/bin/exec/R -q -f clion-test.R
==1435== 
Packages in local lib: assertthat, DBI, dbplyr, dplyr, rlang
Compiling with: 
LTO=-O0
LDFLAGS:=$(LDFLAGS) -fuse-ld=gold

CXX_SOURCES = $(wildcard *.cpp)
DEPFILES = ${CXX_SOURCES:.cpp=.d}

all: ${DEPFILES}
include ${DEPFILES}

PKG_CPPFLAGS := ${PKG_CPPFLAGS} -DPLOGR_ENABLE
PKG_CXXFLAGS := ${PKG_CXXFLAGS} -Winvalid-pch -fpch-preprocess -Wall -Wextra -pedantic -Werror -Wfatal-errors
#PKG_CXXFLAGS := ${PKG_CXXFLAGS} -Winvalid-pch -fpch-preprocess -Wall -Wextra -pedantic -Werror -Wconversion -Wfatal-errors

pch.h.gch: pch.h
	$(CXX) $(ALL_CPPFLAGS) $(ALL_CXXFLAGS) $< -o $@

*.o: pch.h.gch

export CCACHE_SLOPPINESS := pch_defines,time_macros
dplyr-pkg> devtools::load_all()
Loading dplyr
dplyr-pkg> dplyr:::init_logging("INFO")
dplyr-pkg> 
dplyr-pkg> library(modelr)
dplyr-pkg> library(tidyr)
dplyr-pkg> library(purrr)

Attaching package: ‘purrr’

The following objects are masked from ‘package:dplyr’:

    accumulate, accumulate_right, compact, contains, discard, every,
    keep, map, map_chr, map_dbl, map_if, map_int, map_lgl, map2,
    map2_chr, map2_dbl, map2_int, map2_lgl, negate, order_by, pmap,
    reduce, reduce_right, some, transpose

dplyr-pkg> 
dplyr-pkg> gctorture()
dplyr-pkg> 
dplyr-pkg> # minimal example- we're not actually using param, but it
dplyr-pkg> # could be a machine learning tuning parameter
dplyr-pkg> crosses <- mtcars %>%
+   crossv_kfold(k = 5) %>%
+   crossing(param = 1:5)
dplyr-pkg> 
dplyr-pkg> # WARNING: this next line causes a segfault for me:
dplyr-pkg> crosses <- crosses %>%
+   mutate(model = map(train, ~ lm(mpg ~ wt, .)))
dplyr::Result* dplyr::get_handler@141: Looking up hybrid handler for call of type LANGSXP
dplyr::Result* dplyr::get_handler@141: Looking up hybrid handler for call of type LANGSXP
dplyr::Result* dplyr::get_handler@141: Looking up hybrid handler for call of type LANGSXP
dplyr::Result* dplyr::get_handler@141: Looking up hybrid handler for call of type LANGSXP
SEXPREC* dplyr::GroupedHybridEval::eval_with_indices@221: LANGSXP
==1435== Source and destination overlap in memcpy_chk(0x530a404, 0x530a400, 63)
==1435==    at 0x4C34467: __memcpy_chk (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1435==    by 0x4F28AE1: memmove (string3.h:59)
==1435==    by 0x4F28AE1: R_ConciseTraceback.constprop.8 (errors.c:1384)
==1435==    by 0x4F297F6: verrorcall_dflt (errors.c:684)
==1435==    by 0x4F2D79A: errorcall_dflt.constprop.6 (errors.c:712)
==1435==    by 0x4F2D80E: do_dfltStop (errors.c:1734)
==1435==    by 0x4F38344: bcEval (eval.c:5665)
==1435==    by 0x4F4177F: Rf_eval (eval.c:616)
==1435==    by 0x4F433FC: Rf_applyClosure (eval.c:1135)
==1435==    by 0x4F4190C: Rf_eval (eval.c:732)
==1435==    by 0xD4FCF17: dplyr_mutate_impl (RcppExports.cpp:454)
==1435==    by 0x4F098C3: do_dotcall (dotcode.c:1251)
==1435==    by 0x4F41D3E: Rf_eval (eval.c:713)
==1435== 
Error in mutate_impl(.data, dots) : expecting an external pointer
Calls: %>% ... <Anonymous> -> mutate -> mutate.tbl_df -> mutate_impl -> .Call
14: traceback(1)
13: stop(list(message = "expecting an external pointer", call = mutate_impl(.data, 
        dots), cppstack = list(file = "", line = -1L, stack = c("/home/muelleki/git/R/dplyr/dplyr-pkg/src/dplyr.so(+0x18e689) [0xd501689]", 
     ...
12: .Call("dplyr_mutate_impl", PACKAGE = "dplyr", df, dots) at RcppExports.R#134
11: mutate_impl(.data, dots) at tbl-df.r#87
10: mutate.tbl_df(., model = map(train, ~lm(mpg ~ wt, .))) at manip.r#263
9: mutate(., model = map(train, ~lm(mpg ~ wt, .)))
8: function_list[[k]](value)
7: withVisible(function_list[[k]](value)) at freduce.R#12
6: freduce(value, `_function_list`) at pipe.R#28
5: `_fseq`(`_lhs`)
4: eval(expr, envir, enclos)
3: eval(quote(`_fseq`(`_lhs`)), env, env)
2: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env)) at pipe.R#6
1: crosses %>% mutate(model = map(train, ~lm(mpg ~ wt, .)))
==1435== 
==1435== HEAP SUMMARY:
==1435==     in use at exit: 69,981,517 bytes in 27,677 blocks
==1435==   total heap usage: 122,096 allocs, 94,419 frees, 250,310,462 bytes allocated
==1435== 
==1435== LEAK SUMMARY:
==1435==    definitely lost: 0 bytes in 0 blocks
==1435==    indirectly lost: 0 bytes in 0 blocks
==1435==      possibly lost: 0 bytes in 0 blocks
==1435==    still reachable: 69,981,517 bytes in 27,677 blocks
==1435==                       of which reachable via heuristic:
==1435==                         newarray           : 4,264 bytes in 1 blocks
==1435==         suppressed: 0 bytes in 0 blocks
==1435== Rerun with --leak-check=full to see details of leaked memory
==1435== 
==1435== For counts of detected and suppressed errors, rerun with: -v
==1435== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Comment by dgrtwo** [comment]
I've always gotten an error (segmentation fault) when I run it (but I don't follow- I certainly *hoped* it would complete without error!)

**Comment by krlmlr** [comment]
Checking my setup with dplyr 0.5.0 now, if that works, I can bisect. One test takes >10 minutes, so don't hold your breath ;-)

**Comment by krlmlr** [comment]
688911ac5d is the first bad commit, it is related to the fix of #2554. Will fix soon.
